Source/CesiumRuntime/Private/CesiumTextureUtility.cpp

```diff
@@ -1,20 +1,429 @@
+      }
+
+      uint32 mipCount =
+          FMath::Max(1, static_cast<int32>(pImage->mipPositions.size()));
+
+      FTexture2DRHIRef rhiTexture;
+
+      // Copies over mip0, allocates the rest of the mips if needed.
+
+      // RHICreateTexture2D can actually copy over all the mips in one shot,
+      // but it expects a particular memory layout. Might be worth configuring
+      // Cesium Native's mip-map generation to obey a standard memory layout.
+      rhiTexture = RHICreateTexture2D(
+          this->_width,
+          this->_height,
+          this->_format,
+          mipCount,
+          1,
+          textureFlags,
+          createInfo);
+
+      // Copies over rest of the mips
+      for (uint32 i = 1; i < mipCount; ++i) {
+        uint32 DestPitch;
+        void* pDestination =
+            RHILockTexture2D(rhiTexture, i, RLM_WriteOnly, DestPitch, false);
+        CopyMip(pDestination, DestPitch, _format, *pImage, i);
+        RHIUnlockTexture2D(rhiTexture, i, false);
+      }
+
+      this->TextureRHI = rhiTexture;
+      rhiTexture.SafeRelease();
+    }
+
+    RHIUpdateTextureReference(TextureReferenceRHI, this->TextureRHI);
+
+    // TODO: investigate if we can / should do this!
+    // related to: https://github.com/CesiumGS/cesium-unreal/issues/980
+    // this->TextureRHI->DoNoDeferDelete();
+    // this->TextureReferenceRHI->DoNoDeferDelete();
+  }
+
+  virtual void ReleaseRHI() override {
+    RHIUpdateTextureReference(TextureReferenceRHI, nullptr);
+    TextureReferenceRHI.SafeRelease();
+    TextureRHI.SafeRelease();
+
+    FTextureResource::ReleaseRHI();
+  }
+
+private:
+  UTexture* _pTexture;
+  CesiumTextureUtility::CesiumTextureSource _textureSource;
+
+  uint32 _width;
+  uint32 _height;
+  EPixelFormat _format;
+  ESamplerFilter _filter;
+  ESamplerAddressMode _addressX;
+  ESamplerAddressMode _addressY;
+
+  uint32 _platformExtData;
+};
+
+/**
+ * @brief Create an RHI texture on this thread. This requires
+ * GRHISupportsAsyncTextureCreation to be true.
+ *
+ * @param image The CPU image to create on the GPU.
+ * @param format The pixel format of the image.
+ * @param generateMipMaps Whether the RHI texture should have a mipmap.
+ * @return The RHI texture reference.
+ */
+FTexture2DRHIRef CreateRHITexture2D_Async(
+    const CesiumGltf::ImageCesium& image,
+    EPixelFormat format,
+    bool generateMipMaps) {
+  check(GRHISupportsAsyncTextureCreation);
+
+  ETextureCreateFlags textureFlags = TexCreate_ShaderResource | TexCreate_SRGB;
+
+  if (generateMipMaps) {
+    // Here 16 is a generously large (but arbitrary) hard limit for number of
+    // mips.
+    uint32 mipCount = static_cast<uint32>(image.mipPositions.size());
+    if (mipCount > 16) {
+      mipCount = 16;
+    }
+
+    void* mipsData[16];
+    for (size_t i = 0; i < mipCount; ++i) {
+      const CesiumGltf::ImageCesiumMipPosition& mipPos = image.mipPositions[i];
+      mipsData[i] = (void*)&image.pixelData[mipPos.byteOffset];
```

Comment:
I just tried this, a static_cast seems to work in some of the instances where I was using the C-style cast. In other cases where Unreal's API expected a void* for _reading only_, and we had a const reference to buffer or some such - static_cast<void*> isn't allowed since it attempts to cast away the const qualifier. In that case should reinterpret_cast be used? Or const_cast, somehow?

Reply:
Ah ok. The C++ purist way would be to use two casts (static_cast to `const void*` followed by const_cast to `void*`), but the C-style is fine.